storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serde/RemoteLogSegmentMetadataSerde.java

```diff
@@ -0,0 +1,87 @@
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RemoteLogSegmentMetadataSerde implements RemoteLogMetadataSerdes<RemoteLogSegmentMetadata> {
+
+    public Message serialize(RemoteLogSegmentMetadata data) {
+        RemoteLogSegmentMetadataRecord record = new RemoteLogSegmentMetadataRecord()
+                .setRemoteLogSegmentId(
+                        new RemoteLogSegmentMetadataRecord.RemoteLogSegmentIdEntry()
+                                .setTopicIdPartition(new RemoteLogSegmentMetadataRecord.TopicIdPartitionEntry()
+                                        .setId(data.remoteLogSegmentId().topicIdPartition().topicId())
+                                        .setName(data.remoteLogSegmentId().topicIdPartition().topicPartition()
+                                                .topic())
+                                        .setPartition(data.remoteLogSegmentId().topicIdPartition().topicPartition()
+                                                .partition()))
+                                .setId(data.remoteLogSegmentId().id()))
+                .setStartOffset(data.startOffset())
+                .setEndOffset(data.endOffset())
+                .setBrokerId(data.brokerId())
+                .setEventTimestampMs(data.eventTimestampMs())
+                .setMaxTimestampMs(data.maxTimestampMs())
+                .setSegmentSizeInBytes(data.segmentSizeInBytes())
+                .setSegmentLeaderEpochs(data.segmentLeaderEpochs().entrySet().stream()
+                        .map(entry -> new RemoteLogSegmentMetadataRecord.SegmentLeaderEpochEntry()
+                                .setLeaderEpoch(entry.getKey())
+                                .setOffset(entry.getValue())).collect(Collectors.toList()))
+                .setRemoteLogSegmentState(data.state().id());
+
+        return record;
+    }
+
+    @Override
+    public RemoteLogSegmentMetadata deserialize(byte version, ByteBuffer byteBuffer) {
+        RemoteLogSegmentMetadataRecord record = new RemoteLogSegmentMetadataRecord(
+                new ByteBufferAccessor(byteBuffer), version);
+
+        RemoteLogSegmentId remoteLogSegmentId = buildRemoteLogSegmentId(record.remoteLogSegmentId());
+        Map<Integer, Long> segmentLeaderEpochs = new HashMap<>();
+        for (RemoteLogSegmentMetadataRecord.SegmentLeaderEpochEntry segmentLeaderEpoch : record.segmentLeaderEpochs()) {
+            segmentLeaderEpochs.put(segmentLeaderEpoch.leaderEpoch(), segmentLeaderEpoch.offset());
+        }
+        RemoteLogSegmentMetadata remoteLogSegmentMetadata = new RemoteLogSegmentMetadata(remoteLogSegmentId,
+                record.startOffset(), record.endOffset(), record.maxTimestampMs(), record.brokerId(),
+                record.eventTimestampMs(), record.segmentSizeInBytes(), segmentLeaderEpochs);
+        RemoteLogSegmentMetadataUpdate rlsmUpdate = new RemoteLogSegmentMetadataUpdate(remoteLogSegmentId,
```

Comment:
Hmm, why do we need to create and apply an update here?

Reply:
We did not expose `RemoteLogSegmentMetadata` constructor to take state as it is always created with `RemoteLogSegmentState.COPY_SEGMENT_STARTED` and it should always be updated with `RemoteLogSegmentMetadataUpdate`.